EIPS/EIP-8.md

```diff
@@ -0,0 +1,148 @@
+auth-body     = rlp.list(sig, initiator-pubk, initiator-nonce, auth-vsn)
+enc-auth-body = ecies.encrypt(recipient-pubk, auth-body)
+auth-packet   = auth-size || enc-auth-body
+
+ack-vsn       = 4
+ack-size      = size of enc-ack-body, encoded as a big-endian 16-bit integer
+ack-body      = rlp.list(recipient-ephemeral-pubk, recipient-nonce, ack-vsn)
+enc-ack-body  = ecies.encrypt(initiator-pubk, ack-body)
+ack-packet    = ack-size || enc-ack-body
+
+where
+
+X || Y
+    denotes concatenation of X and Y.
+X[:N]
+    denotes an N-byte prefix of X.
+rlp.list(X, Y, Z, ...)
+    denotes recursive encoding of [X, Y, Z, ...] as an RLP list.
+sha3(MESSAGE)
+    is the Keccak256 hash function as used by Ethereum.
+ecies.encrypt(PUBKEY, MESSAGE)
+    is the asymmetric encryption function as used by RLPx.
+```
+
+### Motivation
+
+Changes to the devp2p protocols are hard to deploy because clients running an older
+version will refuse communication if the version number or structure of the hello
+(discovery ping, RLPx handshake) packet does not match local expectations.
+
+Introducing forward-compatibility requirements as part of the Homestead consensus upgrade
+will ensure that all client software in use on the Ethereum network can cope with future
+network protocol upgrades (as long as backwards-compatibility is maintained).
+
+### Rationale
+
+#### Changes to the devp2p Wire Protocol
+
+All clients currently contain statements such as the following:
+
+```python
+# pydevp2p/p2p_protocol.py
+if data['version'] != proto.version:
+    log.debug('incompatible network protocols', peer=proto.peer,
+        expected=proto.version, received=data['version'])
+    return proto.send_disconnect(reason=reasons.incompatibel_p2p_version)
+```
+
+These checks make it impossible to change the version or structure of the initial packet.
+Dropping them enables switching to a newer protocol version: Clients implementing a newer
+version simply send a packet with higher version and possibly additional list elements.
+
+* If such a packet is received by a node with lower version, it will blindly assume that
+  the remote end is backwards-compatible and respond with the old handshake.
+* If the packet is received by a node with equal version, new features of the protocol can
+  be used.
+* If the packet is received by a node with higher version, it can enable
+  backwards-compatibility logic or drop the connection.
+
+#### Changes to the RLPx TCP Handshake
+
+Discussions of the RLPx v5 changes (chunked packets, change to key derivation) have
+faltered in part because the v4 handshake encoding provides only one in-band way to add a
+version number: shortening the random portion of the nonce. Even if the RLPx v5 handshake
+proposal were accepted, future upgrades are hard because the handshake packet is a fixed
+size ECIES ciphertext with known layout.
+
+I propose the following changes to the handshake packets:
+
+* Adding the length of the ciphertext as a plaintext header.
```

Comment:
No plaintext. The problem with this is that the length will *always* be greater than X, and likely equal to X, so it can be inferred that the connection is rlpx. Leaking information (aka plaintext) is not within the design rationale of RLPx or any encrypted protocol.